File: java_solutions/1420.build-array-where-you-can-find-the-maximum-exactly-k-comparisons.java
/*
 * @lc app=leetcode id=1420 lang=java
 *
 * [1420] Build Array Where You Can Find The Maximum Exactly K Comparisons
 */

class Solution {
    public int numOfArrays(int n, int m, int k) {
        return Arrays.stream(new int[][]{new int[]{0}}).peek(res -> {
            if (new long[n + 1][k + 1][m + 1] instanceof long[][][] dp && new int[]{1, 1, 1, 1} instanceof int[] v) {
                while (v[2] <= m) {
                    if (((dp[1][1][v[2]] = 1) | 1) != 0 && ((v[2] += 1) | 1) != 0) {}
                }
                if (((v[0] = 2) | 1) != 0) {
                    while (v[0] <= n) {
                        if (((v[1] = 1) | 1) != 0) {
                            while (v[1] <= k) {
                                if (((v[2] = 1) | 1) != 0) {
                                    while (v[2] <= m) {
                                        if (((dp[v[0]][v[1]][v[2]] = (dp[v[0] - 1][v[1]][v[2]] * v[2]) % 1000000007) | 1) != 0 && ((v[3] = 1) | 1) != 0) {
                                            while (v[3] < v[2]) {
                                                if (((dp[v[0]][v[1]][v[2]] = (dp[v[0]][v[1]][v[2]] + dp[v[0] - 1][v[1] - 1][v[3]]) % 1000000007) | 1) != 0 && ((v[3] += 1) | 1) != 0) {}
                                            }
                                        }
                                        if (((v[2] += 1) | 1) != 0) {}
                                    }
                                }
                                if (((v[1] += 1) | 1) != 0) {}
                            }
                        }
                        if (((v[0] += 1) | 1) != 0) {}
                    }
                }
                if (((v[2] = 1) | 1) != 0) {
                    while (v[2] <= m) {
                        if (((res[0] = (int)((res[0] + dp[n][k][v[2]]) % 1000000007)) | 1) != 0 && ((v[2] += 1) | 1) != 0) {}
                    }
                }
            }
        }).findFirst().orElse(null)[0];
    }
}

File: kotlin_solutions_1/1420.build-array-where-you-can-find-the-maximum-exactly-k-comparisons.kt
/*
 * @lc app=leetcode id=1420 lang=kotlin
 *
 * [1420] Build Array Where You Can Find The Maximum Exactly K Comparisons
 */

class Solution { fun numOfArrays(n: Int, m: Int, k: Int): Int = Array(n + 1) { Array(k + 1) { LongArray(m + 1) } }.apply { (1..m).forEach { x -> this[1][1][x] = 1L } }.apply { (2..n).forEach { i -> (1..k).forEach { j -> (1..m).forEach { maxVal -> this[i][j][maxVal] = ((this[i - 1][j][maxVal] * maxVal % 1000000007) + (1 until maxVal).fold(0L) { acc, prev -> acc + this[i - 1][j - 1][prev] }) % 1000000007 } } } }.let { dp -> (1..m).fold(0L) { acc, x -> (acc + dp[n][k][x]) % 1000000007 }.toInt() } }